data/posts/split-route-modules.md

```diff
@@ -0,0 +1,137 @@
+```
+
+At a framework level, the easiest way for us to solve this would be to force you to author your route in multiple files (`route/clientLoader.ts`, `route/component.tsx`, etc.) — but we really didn't want to give up on the convenience of the Route Module API. The question is, how do we achieve this?
+
+## Splitting the Route Module
+
+What if the React Router Vite plugin could automatically split these route module exports into multiple smaller modules during the production build?
+
+With the `future.unstable_splitRouteModules` flag enabled, this is exactly what happens.
+
+Using our previous example, our singular route module would end up being split into two separate [virtual modules](https://vite.dev/guide/api-plugin#virtual-modules-convention) — one for the client loader and one for the component.
+
+```ts
+// route.tsx?route-chunk=clientLoader
+export async function clientLoader() {
+  return await fetch("https://example.com/api").then((response) =>
+    response.json(),
+  );
+}
+```
+
+```ts
+// route.tsx?route-chunk=main
+import { MassiveComponent } from "~/components";
+
+export default function Component({ loaderData }) {
+  return <MassiveComponent data={loaderData} />;
+}
+```
+
+Since these exports have now been split into separate modules, the React Router Vite plugin can ensure that they can be downloaded independently.
+
+This optimization is even more pronounced when more of the Route Module API is used. For example, when using `clientLoader`, `clientAction` and `HydrateFallback`, the timeline for a single route module during a client-side navigation might look like this:
+
+```
+Get Route Module:     |--~~~++++=======|
+Run clientLoader:                      |-----|
+Render:                                      |-|
+```
+
+This would instead be optimized to the following:
+
+```
+Get clientLoader:     |--|
+Get clientAction:     |~~~|
+Get HydrateFallback:  SKIPPED
+Get Component:        |=======|
+Run clientLoader:        |-----|
+Render:                        |-|
+```
+
+This looks much better! As before, the client loader doesn't need to wait for the component to download — and now it doesn't need to wait for the `clientAction` or `HydrateFallback` exports to download either. In fact, it doesn't even need to download the `HydrateFallback` export at all during client navigations since it's only ever used on the initial page load.
+
+## Try it out
+
+This feature will be enabled by default in a future release, but you can try it out today by setting the `future.unstable_splitRouteModules` flag in your React Router config:
+
+```ts
+// react-router.config.ts
+import type { Config } from "@react-router/dev/config";
+
+export default {
+  future: {
+    unstable_splitRouteModules: true,
+  },
+} satisfies Config;
+```
+
+If your project is especially performance sensitive, you can set `future.unstable_splitRouteModules` to `"enforce"`. This will break the build if any route module cannot be split.
```

Comment:
This feels like it's missing some context - should we include a small section with the scenarios that prohibit splitting and maybe a small code example?

Reply:
Good point, I could expand on this.